File: Aardvark-WPILib/SolenoidBase.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) FIRST 2008. All Rights Reserved.							  */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in $(WIND_BASE)/WPILib.  */
/*----------------------------------------------------------------------------*/

#include "SolenoidBase.h"

// Needs to be global since the protected resource spans all Solenoid objects.
ReentrantSemaphore SolenoidBase::m_semaphore;
Resource *SolenoidBase::m_allocated = NULL;

tSolenoid *SolenoidBase::m_fpgaSolenoidModule = NULL;
uint32_t SolenoidBase::m_refCount = 0;


/**
 * Constructor
 * 
 * @param moduleNumber The solenoid module (1 or 2).
 */
SolenoidBase::SolenoidBase(uint8_t moduleNumber)
	: m_moduleNumber (moduleNumber)
{
	Synchronized sync(m_semaphore);
	m_refCount++;
	if (m_refCount == 1)
	{
		tRioStatusCode localStatus = NiFpga_Status_Success;
		m_fpgaSolenoidModule = tSolenoid::create(&localStatus);
		wpi_setError(localStatus);
	}
}

/**
 * Destructor.
 */
SolenoidBase::~SolenoidBase()
{
	Synchronized sync(m_semaphore);
	if (CheckSolenoidModule(m_moduleNumber))
	{
		if (m_refCount == 1)
		{
			delete m_fpgaSolenoidModule;
			m_fpgaSolenoidModule = NULL;
		}
		m_refCount--;
	}
}

/**
 * Set the value of a solenoid.
 * 
 * @param value The value you want to set on the module.
 * @param mask The channels you want to be affected.
 */
void SolenoidBase::Set(uint8_t value, uint8_t mask)
{
	tRioStatusCode localStatus = NiFpga_Status_Success;
	if (CheckSolenoidModule(m_moduleNumber))
	{
		Synchronized sync(m_semaphore);
		uint8_t currentValue = m_fpgaSolenoidModule->readDO7_0(m_moduleNumber - 1, &localStatus);
		// Zero out the values to change
		currentValue = currentValue & ~mask;
		currentValue = currentValue | (value & mask);
		m_fpgaSolenoidModule->writeDO7_0(m_moduleNumber - 1, currentValue, &localStatus);
	}
	wpi_setError(localStatus);
}

/**
 * Read all 8 solenoids as a single byte
 * 
 * @return The current value of all 8 solenoids on the module.
 */
uint8_t SolenoidBase::GetAll()
{
	if (CheckSolenoidModule(m_moduleNumber))
	{
		tRioStatusCode localStatus = NiFpga_Status_Success;
		uint8_t solenoids = m_fpgaSolenoidModule->readDO7_0(m_moduleNumber - 1, &localStatus);
		wpi_setError(localStatus);
		return solenoids;
	}
	return 0;
}

File: originalCode/Aardvark-WPILib-Java/src/edu/wpi/first/wpilibj/SolenoidBase.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) FIRST 2008-2012. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj;

import edu.wpi.first.wpilibj.fpga.tSolenoid;
import edu.wpi.first.wpilibj.parsing.IDeviceController;

/**
 * SolenoidBase class is the common base class for the Solenoid and
 * DoubleSolenoid classes.
 */
public abstract class SolenoidBase extends SensorBase implements IDeviceController {

    protected int m_moduleNumber; ///< The number of the solenoid module being used.
    protected static Resource m_allocated = new Resource(tSolenoid.kDO7_0_NumElements * kSolenoidChannels);

    private static tSolenoid m_fpgaSolenoidModule; ///< FPGA Solenoid Module object.
    private static int m_refCount; ///< Reference count for the chip object.

    /**
     * Constructor.
     *
     * @param moduleNumber The number of the solenoid module to use.
     */
    public SolenoidBase(final int moduleNumber) {
        m_moduleNumber = moduleNumber;
        checkSolenoidModule(m_moduleNumber);

        m_refCount++;
        if (m_refCount == 1) {
            m_fpgaSolenoidModule = new tSolenoid();
        }
    }

    /**
     * Destructor.
     */
    public synchronized void free() {
        if (m_refCount == 1) {
            m_fpgaSolenoidModule.Release();
            m_fpgaSolenoidModule = null;
        }
        m_refCount--;
    }

    /**
     * Set the value of a solenoid.
     *
     * @param value The value you want to set on the module.
     * @param mask The channels you want to be affected.
     */
    protected synchronized void set(int value, int mask) {
        byte currentValue = (byte)tSolenoid.readDO7_0(m_moduleNumber - 1);
        // Zero out the values to change
        currentValue = (byte)(currentValue & ~mask);
        currentValue = (byte)(currentValue | (value & mask));
        tSolenoid.writeDO7_0(m_moduleNumber - 1, currentValue);
    }

    /**
     * Read all 8 solenoids from the module used by this solenoid as a single byte
     *
     * @return The current value of all 8 solenoids on this module.
     */
    public byte getAll() {
        return (byte)tSolenoid.readDO7_0(m_moduleNumber - 1);
    }

    /**
     * Read all 8 solenoids in the default solenoid module as a single byte
     *
     * @return The current value of all 8 solenoids on the default module.
     */
    public static byte getAllFromDefaultModule() {
        return getAllFromModule(getDefaultSolenoidModule());
    }

    /**
     * Read all 8 solenoids in the specified solenoid module as a single byte
     *
     * @return The current value of all 8 solenoids on the specified module.
     */
    public static byte getAllFromModule(int moduleNumber) {
        checkSolenoidModule(moduleNumber);
        return (byte)tSolenoid.readDO7_0(moduleNumber - 1);
    }
}
